Various updates to mortar contact

recuero opened the pull request:
Various changes to mortar contact. Including:

- Incorporate `sparsity_trim(tolerance)` into `ADMortarLagrangeConstraint` as a handle to remove insignificant derivatives. I'd probably only use this in extreme occasions in which I suspect there are terms in AD objects that do not correspond to meaningful Jacobian entries --> Uncomment line when MetaPhysicL patch makes it.
- Avoid heap allocation when holding onto tangent vectors
- Allow the contact action to not generate lower-d domains. Useful for restarts.
- Allow use of dynamic mortar contact constraints from the action
- Alex's update to skip_partitioning/allow_renumbering in `FileMeshGenerator`

Refs #17495, but changes are general.

recuero committed:
Reduce heap memory allocation/deallocation with frictional mortar

When profiling the code, the amount of memory allocation/deallocation
pertaining to standard vectors that hold tangent vectors stood out (1.5GB)
for a mid-size problem. It was by far the largest hit setting aside Jacobian
entry allocations. For this reason, this commit replaces the std::vector with
the semidynamic vector that Daniel created based on the MetaPhysicL library.
 This change should have no effect other than better memory management.

recuero committed:
Allow user to disable mesh generation in mortar contact action (#17495)

This can come in handy when restarting and wanting to use the existing
lower dimensional blocks.

recuero committed:
Touch up automatic mortar generation internal tolerances

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/998879/) on f947d2d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20665/site)

This comment will be updated on new commits.

recuero committed:
Fixes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/998990/) on 28ea5d3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20665/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20665/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c90b1f04c38b5fa8c35d8b16e7ca4ae7c7ad7115`

recuero committed:
Enable contact mechanics action to work with dynamic mortar constraints

lindsayad committed:
Apply skip_partitioning and allow_renumbering to FileMeshGenerator

recuero committed:
Check dynamic frictional action results against non-action intput.

Also add a missing Exodus mesh file/add capture_tolerance to dynamic
frictional constraint.

recuero commented:
The addition of `sparsity_trim(tol)` to `ADMortarLagrangeConstraint` appears to generate a type mismatch caught in debug mode from the `sparsity_trim` routine when called from MOOSE. Trimming the derivatives this way, even in amounts very close to zero, also causes one no_adaptivity VCP test to fail, which may rely on indices that are deleted. So, for now, I am not adding this type of tolerance-based derivative trimming.

recuero committed:
Test fixes

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1000703/) on e4d0e2a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>c90b1f</th>
      <th colspan=4>#20665 e4d0e2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.56%</td>
      <td>82.56%</td>
      <td>+0.00%</td>
      <td>58.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73757</td>
      <td>73763</td>
      <td>+6</td>
      <td>7</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15581</td>
      <td>15580</td>
      <td>-1</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20665/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20665/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>c90b1f</th>
      <th colspan=4>#20665 e4d0e2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>88.95%</td>
      <td>-0.06%</td>
      <td>87.50%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2917</td>
      <td>2938</td>
      <td>+21</td>
      <td>42</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>360</td>
      <td>365</td>
      <td>+5</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20665/coverage_diff/contact.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20665/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20665/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 58.33% is less than the suggested 90.0%
- `contact` new line coverage rate 87.50% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Increase coverage in contact module

moosebuild commented:
Job [Apple Silicon Stack](https://civet.inl.gov/job/1000918/) on 996645c : invalidated by @recuero

moosebuild commented:
Job [Apple Silicon Stack](https://civet.inl.gov/job/1000918/) on 996645c : invalidated by @recuero

moosebuild commented:
Job [Apple Silicon Stack](https://civet.inl.gov/job/1000918/) on 996645c : invalidated by @recuero

lindsayad commented:
Any failures on Apple may not be your fault. See https://github.com/idaholab/moose/pull/20669#issuecomment-1085965026

recuero committed:
Merge branch 'next' into misc_mortar

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -182,7 +183,8 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
    * @return The nodal tangents associated with the provided \p secondary_elem
    */
-  std::array<std::vector<Point>, 2> getNodalTangents(const Elem & secondary_elem) const;
+  std::array<MooseUtils::SemidynamicVector<Point, 10>, 2>
```
I'm curious how we arrived at 10? 9 (not very different) seems like a good number because QUAD9 should be the "biggest" face we ever have

recuero replied:
Sure, as of now 9 should be sufficient.

recuero replied:
Made it 9.

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -1739,7 +1740,10 @@ AutomaticMortarGeneration::projectSecondaryNodesSinglePair(
+
+            // Increasing the tolerance 5 times in this if statement to avoid missing nodes due to
+            // the mismatch between sizes of primary and secondary surfaces. Still, it can happen.
+            if (std::abs(std::abs(xi2) - 1.) < _xi_tolerance * 5.0)
```
should we just make `_xi_tolerance` 5 times bigger?

recuero replied:
Hmm I don't think that's the issue. We are projecting on one side first and the other side second. If we make the tolerance bigger and remove the (5) factor we are going to continue to miss the second projection and fall into that message. What makes this modification to not fall in the exception is that we are projecting on one side more `xi` than in the other. There should be a better way of doing this by using actual distances and not parametric coordinates. But I believe making the tolerance uniformly larger or smaller won't do the trick here.

lindsayad replied:
Ok. I would put that logic in the comment because it helps me understand

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -2033,8 +2038,13 @@ AutomaticMortarGeneration::projectPrimaryNodesSinglePair(
               // Special case: xi1=+/-1.
               // We shouldn't get here, because this primary node should already
               // have been mapped during the project_secondary_nodes() routine.
-              throw MooseException("We should never get here, aligned primary nodes should already "
-                                   "have been mapped.");
+              // I think there is a chance we get here since the tolerances are applied to
+              // the xi coordinate and that value may be different on a primary element and a
+              // secondary element since they may have different sizes.
```
Maybe something like

```
// It is unlikely that we get here, because this primary node should already
// have been mapped during the project_secondary_nodes() routine, but
// there is still a chance since the tolerances are applied to
// the xi coordinate and that value may be different on a primary element and a
// secondary element since they may have different sizes.
```

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -2033,8 +2038,13 @@ AutomaticMortarGeneration::projectPrimaryNodesSinglePair(
+              // I think there is a chance we get here since the tolerances are applied to
+              // the xi coordinate and that value may be different on a primary element and a
+              // secondary element since they may have different sizes.
+
```
```suggestion
```

lindsayad commented on modules/contact/include/constraints/TangentialMortarMechanicalContact.h:
```diff
@@ -10,6 +10,7 @@
 #pragma once
 
 #include "ADMortarLagrangeConstraint.h"
+#include "MooseUtils.h"
```
```suggestion
```

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -181,8 +205,8 @@ ContactAction::ContactAction(const InputParameters & params)
       mooseDeprecated(
-          "Use of legacy mortar contact approach is deprecated and will be removed by December "
-          "2021. Instead, select the default option based on weighted quantities and dual bases");
+          "Use of legacy mortar contact approach is deprecated and will be removed by September "
+          "2022. Instead, select the default option based on weighted quantities and dual bases");
```
😆 I occasionally run across deprecation messages that I put in that said I would remove something a long time ago ... and clearly haven't

recuero replied:
I am planning to do this by September, just postpone it over and over given that there are other more pressing items to tackle, but definitely plan to do it!

lindsayad commented on modules/contact/src/actions/ContactAction.C:
```diff
@@ -471,8 +509,10 @@ ContactAction::addMortarContact()
   {
-    // Don't do mesh generators when recovering.
-    if (!(_app.isRecovering() && _app.isUltimateMaster()) && !_app.masterMesh())
+    // Don't do mesh generators when  or just don't want it to be generated because the mesh is
+    // already in the system (e.g. restart).
```
```suggestion
    // Don't do mesh generators when recovering or when the user has requested for us not to (presumably
    // because the lower-dimensional blocks are already in the mesh due to manual addition or because we
    // are restarting)
```

lindsayad reviewed (changes requested):
Partway through reviewing

lindsayad commented on modules/contact/src/constraints/ADMortarLagrangeConstraint.C:
```diff
@@ -159,7 +159,10 @@ ADMortarLagrangeConstraint::computeJacobian(Moose::MortarType mortar_type)
       residuals_lower[index_lower] += _JxW_msm[_qp] * _coord[_qp] * computeQpResidual(mortar_type);
 
-      // Get rid of derivatives that we assume won't count (tolerance prescribed by user)
+      // TODO: Get rid of derivatives that we assume won't count (tolerance prescribed by user)
+      // This can cause zero diagonal terms with the variable condensation preconditioner when the
+      // no adaptivity option is used (dofs are not checked).
+
```
This is unfortunate. Would it make sense to make a boolean user parameter that defaults to true to allow trimming, and that should be turned to false when running the VCP? Is the VCP setup through the `ContactAction` or no?

lindsayad replied:
Actually better yet would be during `initialSetup` to just check (`dynamic_cast`) the preconditioner type and don't trip if it's VCP. Then no additional user parameter required

recuero replied:
The VCP cannot be set up via contact action yet. It is still undergoing some work to make it more general for 3D applications/non-trivial geometries.

I can implement that strategy and set that flag to false for that particular test. Since we are removing Jacobian entries I am a bit hesitant to do this now should this have unintended consequences that we are not even catching in MOOSE testing. Anyway, I can give it a try and monitor it. The good part of this is that only the "no adaptivity" test is affected, which wouldn't be the most general case for VCP.   

But I am also getting a type mismatch when calling that routine from MOOSE: https://github.com/idaholab/moose/pull/20665#issuecomment-1085001008

lindsayad replied:
Can you paste the compiler error?

recuero replied:
https://civet.inl.gov/job/1010054/

lindsayad commented on modules/contact/test/tests/mortar_dynamics/tests:
```diff
@@ -16,6 +16,45 @@
+               'Executioner/end_time=0.3'
+    min_ad_size = '50'
+    ad_indexing_type = 'global'
+    requirement = 'Contact module shall generate an error if dynamics is not specifically requested '
```
```suggestion
    requirement = 'The contact module shall generate an error if dynamics is not specifically requested '
```

lindsayad commented on modules/contact/test/tests/mortar_dynamics/tests:
```diff
@@ -16,6 +16,45 @@
+               'Executioner/end_time=0.3'
+    min_ad_size = '50'
+    ad_indexing_type = 'global'
+    requirement = 'Contact module shall generate an error if dynamics is not specifically requested '
```
```suggestion
    requirement = 'The contact module shall generate an error if dynamics is not specifically requested '
```

recuero committed:
Address Alex's review

recuero commented:
@lindsayad I think your comments should be addressed. But I am still getting the type mismatch in debug mode down in the MetaPhysicL library.

lindsayad commented:
It looks good. Thanks for sharing the MetaPhysicL error. Feel free to comment that line again as it will take a MetaPhysicL update to make that go away and I doubt you want to wait that long to get this in

recuero committed:
Do not trim until MetaPhysicL patch makes it to MOOSE.

We will be able to revert this commit with the next update. Then, the
residual vector in ADMortarLagrangeConstraint will be trimmed based on a
default or user-prescribed value if the variable condensation preconditioner
isn't used.
